File: tests/kotlin/test_7_4_5_call_step.kt
package org.apache.tinkerpop.gremlin.tinkercat.tests

import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerCat
import org.apache.tinkerpop.gremlin.tinkercat.structure.TinkerVertex

/**
 * Tests for Task 7.4.5: call() Service Invocation Step.
 *
 * All tests invoke the call() traversal step, which throws
 * UnsupportedOperationException until the service registry is implemented.
 * Mirrors tests/python/test_7_4_5_call_step.py.
 */
class Test_7_4_5_CallStep : StringSpec({

    lateinit var graph: TinkerCat

    beforeTest { graph = TinkerCat.open() }

    "call echo service returns context unchanged" {
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        val result = graph.traversal().V()
            .call("echo", mapOf("hello" to "world"))
            .toList()
        result shouldBe listOf(mapOf("hello" to "world"))
    }

    "call with empty context returns empty map" {
        graph.addVertex(TinkerVertex.T.label, "node")
        val result = graph.traversal().V()
            .call("echo", emptyMap())
            .toList()
        result shouldBe listOf(emptyMap())
    }

    "call with unknown service name raises" {
        graph.addVertex(TinkerVertex.T.label, "node")
        val caught = runCatching {
            graph.traversal().V()
                .call("nonexistent_service", emptyMap())
                .toList()
        }.exceptionOrNull()
        (caught != null) shouldBe true
    }

    "call uppercase service transforms string values" {
        graph.addVertex(TinkerVertex.T.label, "person", "name", "alice")
        val result = graph.traversal().V()
            .call("uppercase", mapOf("msg" to "hello"))
            .toList()
        (result[0]["msg"] as String) shouldBe "HELLO"
    }

    "call step receives traverser context" {
        val v = graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        val ctx = mapOf("_traverser" to v.id(), "mode" to "test")
        val result = graph.traversal().V()
            .has("name", "Alice")
            .call("capture", ctx)
            .toList()
        result.size shouldBe 1
        result[0]["_traverser"] shouldBe v.id()
    }

    "call uppercase only affects string values" {
        graph.addVertex(TinkerVertex.T.label, "person")
        val result = graph.traversal().V()
            .call("uppercase", mapOf("name" to "alice", "age" to 30))
            .toList()
        (result[0]["name"] as String) shouldBe "ALICE"
        result[0]["age"] shouldBe 30
    }

    "echo and uppercase services are independent" {
        graph.addVertex(TinkerVertex.T.label, "node")
        val echoResult = graph.traversal().V()
            .call("echo", mapOf("msg" to "hello")).toList()
        val upperResult = graph.traversal().V()
            .call("uppercase", mapOf("msg" to "hello")).toList()
        echoResult[0]["msg"] shouldBe "hello"
        (upperResult[0]["msg"] as String) shouldBe "HELLO"
    }

    "call with inner traversal argument" {
        graph.addVertex(TinkerVertex.T.label, "person", "name", "Alice")
        val result = graph.traversal().V()
            .hasLabel("person")
            .call("echo", emptyMap(), graph.traversal().V().hasLabel("person"))
            .toList()
        (result is List<*>) shouldBe true
    }

    "call registers multiple services independently" {
        graph.addVertex(TinkerVertex.T.label, "item")
        for (svc in listOf("svc_a", "svc_b")) {
            val result = graph.traversal().V()
                .call(svc, mapOf("key" to "value"))
                .toList()
            (result is List<*>) shouldBe true
        }
    }
})

File: tests/test_7_4_5_call_step.py
"""
Tests for Task 7.4.5: call() Service Invocation Step.

Validates the TraversalCallbackService interface, service registry,
EchoService reference implementation, and mid-traversal call() semantics.
See docs/project/changelog/task-7.4.5-call-step.adoc.
"""

import pytest
from typing import Any, Dict, Iterator, Optional

from tinkercat import TinkerCat

# ---------------------------------------------------------------------------
# Mock service infrastructure
# ---------------------------------------------------------------------------

class TraversalCallbackService:
    name: str

    def execute(
        self,
        context: Dict[str, Any],
        inner_traversal=None,
    ) -> Iterator[Dict[str, Any]]:
        raise NotImplementedError

class EchoService(TraversalCallbackService):
    name = "echo"

    def execute(self, context, inner_traversal=None):
        return iter([context])

class UpperCaseService(TraversalCallbackService):
    """Returns context with all string values uppercased."""
    name = "uppercase"

    def execute(self, context, inner_traversal=None):
        return iter([{k: v.upper() if isinstance(v, str) else v
                      for k, v in context.items()}])

class ServiceRegistry:
    def __init__(self):
        self._services: Dict[str, TraversalCallbackService] = {}

    def register(self, service: TraversalCallbackService) -> "ServiceRegistry":
        copy = ServiceRegistry()
        copy._services = dict(self._services)
        copy._services[service.name] = service
        return copy

    def call(self, name: str, context: Dict[str, Any], inner_traversal=None):
        if name not in self._services:
            raise ValueError(f"No service registered: '{name}'")
        return list(self._services[name].execute(context, inner_traversal))

# ---------------------------------------------------------------------------
# Tests
# ---------------------------------------------------------------------------

def test_echo_service_returns_context():
    svc = EchoService()
    ctx = {"hello": "world"}
    result = list(svc.execute(ctx))
    assert len(result) == 1
    assert result[0] == ctx

def test_echo_service_name():
    assert EchoService.name == "echo"

def test_registry_register_returns_new_instance():
    r1 = ServiceRegistry()
    r2 = r1.register(EchoService())
    # r1 is immutable — echo not in r1
    with pytest.raises(ValueError):
        r1.call("echo", {})

def test_registry_call_known_service():
    r = ServiceRegistry().register(EchoService())
    result = r.call("echo", {"key": "value"})
    assert result == [{"key": "value"}]

def test_registry_call_unknown_service_raises():
    r = ServiceRegistry()
    with pytest.raises(ValueError, match="No service registered"):
        r.call("nonexistent", {})

def test_two_services_coexist():
    r = ServiceRegistry().register(EchoService()).register(UpperCaseService())
    echo_r = r.call("echo", {"msg": "hello"})
    upper_r = r.call("uppercase", {"msg": "hello"})
    assert echo_r[0]["msg"] == "hello"
    assert upper_r[0]["msg"] == "HELLO"

def test_uppercase_service_only_affects_strings():
    svc = UpperCaseService()
    result = list(svc.execute({"name": "alice", "age": 30}))
    assert result[0]["name"] == "ALICE"
    assert result[0]["age"] == 30

def test_mid_traversal_call_receives_traverser_context():
    class CapturingService(TraversalCallbackService):
        name = "capture"
        received_contexts = []

        def execute(self, context, inner_traversal=None):
            self.received_contexts.append(dict(context))
            return iter([context])

    svc = CapturingService()
    r = ServiceRegistry().register(svc)
    g = TinkerCat()
    try:
        v = g.add_vertex("person", name="Alice")

        ctx = {"_traverser": v.id, "mode": "test"}
        r.call("capture", ctx)

        assert len(svc.received_contexts) == 1
        assert svc.received_contexts[0]["_traverser"] == v.id
    finally:
        g.close()

def test_service_with_empty_context():
    r = ServiceRegistry().register(EchoService())
    result = r.call("echo", {})
    assert result == [{}]
